sandbox/host/src/sandbox/wasmer_backend.rs

```diff
@@ -382,6 +424,124 @@ fn dispatch_function(
     })
 }
 
+fn dispatch_function2(
+    supervisor_func_index: SupervisorFuncIndex,
+    store: &wasmer::Store,
+    func_ty: &wasmer::FunctionType,
+) -> wasmer::Function {
+    wasmer::Function::new_with_env(store, func_ty, Env::default(), move |env, params| {
+        SandboxContextStore::with(|sandbox_context| {
+            let gas = env
+                .gas
+                .as_ref()
+                .ok_or_else(|| RuntimeError::new("gas global should be set"))?;
+
+            // Serialize arguments into a byte vector.
+            let invoke_args_data = [gas.get()]
+                .iter()
+                .chain(params.iter())
+                .map(|val| match val {
+                    wasmer::Val::I32(val) => Ok(Value::I32(*val)),
+                    wasmer::Val::I64(val) => Ok(Value::I64(*val)),
+                    wasmer::Val::F32(val) => Ok(Value::F32(f32::to_bits(*val))),
+                    wasmer::Val::F64(val) => Ok(Value::F64(f64::to_bits(*val))),
+                    _ => Err(RuntimeError::new(format!(
+                        "Unsupported function argument: {:?}",
+                        val
+                    ))),
+                })
+                .collect::<std::result::Result<Vec<_>, _>>()?
+                .encode();
+
+            // Move serialized arguments inside the memory, invoke dispatch thunk and
+            // then free allocated memory.
+            let invoke_args_len = invoke_args_data.len() as WordSize;
```

Comment:
Can't we unify common parts of the fn v1 and v2?

Reply:
Yep, we are able to some degree however there is no sense for it from my point of view. V1 is kept just for compatibility with old runtimes and it wouldn't changed anymore.

Reply:
Anyway will add TODO